File: csharp/layers/layer_tree/layer_tree.cs
// ============================================================================
//  layer_tree -- C# port of examples\Vb6\layers\layer_tree\layer_tree.bas
//  Creates three optional-content groups (layers) arranged in a display tree with
//  a titled group, and adds text (with a web link) and an image to the layers.
// ============================================================================
using System;
using System.IO;
using LumasPdfSdk;

class LayerTree
{
    const uint clBlue = 0xFF0000;
    const uint clBlack = 0x0;

    static TErrorProc _err = PDFError;

    static int PDFError(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine(errMessage);
        return 0;
    }

    static void Main()
    {
        string exeDir = AppDomain.CurrentDomain.BaseDirectory;

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfCreateNewPDFW(pdf, "");   // output file opened later

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        // Disable color key masking for images
        LumasPdf.pdfSetUseTransparency(pdf, false);

        // Create three layers
        int oc1 = LumasPdf.pdfCreateOCGW(pdf, "All", false, true, LumasPdfConsts.oiAll);
        int oc2 = LumasPdf.pdfCreateOCGW(pdf, "Text and Annotations", false, true, LumasPdfConsts.oiAll);
        int oc3 = LumasPdf.pdfCreateOCGW(pdf, "Images", false, true, LumasPdfConsts.oiAll);

        IntPtr root = LumasPdf.pdfAddLayerToDisplTreeW(pdf, IntPtr.Zero, oc1, "A layer group with a title");
        IntPtr grp = LumasPdf.pdfAddLayerToDisplTreeW(pdf, root, -1, "");
        LumasPdf.pdfAddLayerToDisplTreeW(pdf, grp, oc2, "");
        LumasPdf.pdfAddLayerToDisplTreeW(pdf, grp, oc3, "");

        LumasPdf.pdfAppend(pdf);
        // The main layer controls the visibility of all three layers in this example.
        LumasPdf.pdfBeginLayer(pdf, (uint)oc1);
        LumasPdf.pdfBeginLayer(pdf, (uint)oc2);
        LumasPdf.pdfSetFontW(pdf, "Helvetica", LumasPdfConsts.fsRegular, 12.0, false, TCodepage.cp1252);
        string someText = "Some text with a link!!!";
        LumasPdf.pdfSetFillColor(pdf, clBlue);
        LumasPdf.pdfWriteTextW(pdf, 50.0, 50.0, someText);
        double tw = LumasPdf.pdfGetTextWidthW(pdf, someText);
        LumasPdf.pdfSetBorderStyle(pdf, (int)TBorderStyle.bsUnderline);
        LumasPdf.pdfSetStrokeColor(pdf, clBlue);
        int annot = LumasPdf.pdfWebLinkW(pdf, 50.0, 51.0, tw, 12.0, "www.lumaspdf.com");

        uint[] ocArray = new uint[] { (uint)oc1, (uint)oc2 };
        int ocmd = LumasPdf.pdfCreateOCMD(pdf, TOCVisibility.ovAllOn, ocArray, 2);
        LumasPdf.pdfAddObjectToLayer(pdf, (uint)ocmd, TOCObject.ooAnnotation, (uint)annot);
        LumasPdf.pdfEndLayer(pdf);

        LumasPdf.pdfBeginLayer(pdf, (uint)oc3);
        LumasPdf.pdfInsertImageExW(pdf, 50.0, 70.0, 300.0, 200.0,
            Path.Combine(exeDir, "margarita-102572_640.jpg"), 1);
        LumasPdf.pdfEndLayer(pdf);
        LumasPdf.pdfEndLayer(pdf);

        LumasPdf.pdfSetFillColor(pdf, clBlack);
        LumasPdf.pdfWriteTextW(pdf, 50.0, 300.0, "This text is not part of a layer!");
        LumasPdf.pdfEndPage(pdf);

        LumasPdf.pdfSetPageMode(pdf, (int)TPageMode.pmUseOC);

        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            string outFile = Path.Combine(exeDir, "out.pdf");
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
            if (LumasPdf.pdfCloseFile(pdf))
                Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        }
        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: c/layers/layer_tree/layer_tree.c
/* layer_tree -- C port of examples\Vb6\layers\layer_tree\layer_tree.bas */
#include <stdio.h>
#include <string.h>
#include "lumaspdf.h"

#define clBlue  0xFF0000u
#define clBlack 0x0u

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    printf("%s\n", ErrMessage);
    return 0;
}

static void exedir(const char* a0, char* out, size_t n)
{
    char* s;
    strncpy(out, a0, n - 1); out[n - 1] = 0;
    s = strrchr(out, '\\'); if (!s) s = strrchr(out, '/');
    if (s) *s = 0; else strcpy(out, ".");
}

int main(int argc, char** argv)
{
    PPDF pdf;
    SI32 annot, ocmd, oc1, oc2, oc3;
    void* root; void* grp;
    double tw;
    char dir[1024], outFile[1100];
    const char* someText = "Some text with a link!!!";
    UI32 ocArray[2];

    exedir(argv[0], dir, sizeof(dir));
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);
    pdfSetUseTransparency(pdf, 0);

    oc1 = pdfCreateOCGA(pdf, "All", 0, 1, oiAll);
    oc2 = pdfCreateOCGA(pdf, "Text and Annotations", 0, 1, oiAll);
    oc3 = pdfCreateOCGA(pdf, "Images", 0, 1, oiAll);

    root = pdfAddLayerToDisplTreeA(pdf, 0, oc1, "A layer group with a title");
    grp = pdfAddLayerToDisplTreeA(pdf, root, -1, "");
    pdfAddLayerToDisplTreeA(pdf, grp, oc2, "");
    pdfAddLayerToDisplTreeA(pdf, grp, oc3, "");

    pdfAppend(pdf);
    pdfBeginLayer(pdf, oc1);
    pdfBeginLayer(pdf, oc2);
        pdfSetFontA(pdf, "Helvetica", fsRegular, 12.0, 0, cp1252);
        pdfSetFillColor(pdf, clBlue);
        pdfWriteTextA(pdf, 50.0, 50.0, someText);
        tw = pdfGetTextWidthA(pdf, someText);
        pdfSetBorderStyle(pdf, bsUnderline);
        pdfSetStrokeColor(pdf, clBlue);
        annot = pdfWebLinkA(pdf, 50.0, 51.0, tw, 12.0, "www.lumaspdf.com");

        ocArray[0] = (UI32)oc1;
        ocArray[1] = (UI32)oc2;
        ocmd = pdfCreateOCMD(pdf, ovAllOn, ocArray, 2);
        pdfAddObjectToLayer(pdf, ocmd, ooAnnotation, annot);
    pdfEndLayer(pdf);

    pdfBeginLayer(pdf, oc3);
        pdfInsertImageExA(pdf, 50.0, 70.0, 300.0, 200.0, "../../../test_files/images/margarita-102572_640.jpg", 1);
    pdfEndLayer(pdf);
    pdfEndLayer(pdf);

    pdfSetFillColor(pdf, clBlack);
    pdfWriteTextA(pdf, 50.0, 300.0, "This text is not part of a layer!");
    pdfEndPage(pdf);

    pdfSetPageMode(pdf, pmUseOC);

    if (pdfHaveOpenDoc(pdf) != 0) {
        sprintf(outFile, "%s\\out.pdf", dir);
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 0; }
        if (pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile);
    }
    pdfDeletePDF(pdf);
    return 0;
}
